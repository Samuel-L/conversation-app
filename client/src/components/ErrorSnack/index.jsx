import React from 'react';
import PropTypes from 'prop-types';

import { withStyles } from '@material-ui/core/styles';
import Snackbar from '@material-ui/core/Snackbar';
import SnackbarContent from '@material-ui/core/SnackbarContent';
import ErrorIcon from '@material-ui/icons/Error';
import IconButton from '@material-ui/core/IconButton';
import CloseIcon from '@material-ui/icons/Close';

const styles = theme => ({
  error: {
    backgroundColor: theme.palette.error.dark,
  },
  icon: {
    fontSize: 20,
    marginRight: theme.spacing.unit,
  },
  message: {
    display: 'flex',
    alignItems: 'center',
  },
});

const ErrorSnack = (props) => {
  const { classes, message } = props;

  return (
    <Snackbar
      anchorOrigin={{ vertical: 'bottom', horizontal: 'left' }}
      open={props.error}
      onClose={props.handleClose}
      autoHideDuration={5000}
    >
      <SnackbarContent
        className={classes.error}
        aria-describedby="error-snackbar"
        message={
          <span id="error-snackbar" className={classes.message}>
            <ErrorIcon className={classes.icon} />
            { message }
          </span>
        }
        action={
          <IconButton
            key="close"
            aria-label="Close"
            color="inherit"
            className={classes.close}
            onClick={props.handleClose}
          >
            <CloseIcon className={classes.icon} />
          </IconButton>
        }
      />
    </Snackbar>
  );
};

ErrorSnack.propTypes = {
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
  message: PropTypes.string.isRequired,
  error: PropTypes.bool.isRequired,
  handleClose: PropTypes.func.isRequired,
};

export default withStyles(styles)(ErrorSnack);
