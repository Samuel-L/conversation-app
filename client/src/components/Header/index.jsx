import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import compose from 'recompose/compose';

import { withStyles } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';

import { siteTitle } from '../../utils/helpers';
import { logout } from '../../redux-modules/auth';

const styles = {
  root: {
    flexGrow: 1,
  },
  flex: {
    flex: 1,
  },
};

export const Header = (props) => {
  const { classes, isLoggedIn } = props;

  return (
    <div className={classes.root}>
      <AppBar position="static">
        <Toolbar>
          <Typography variant="title" color="inherit" className={classes.flex}>
            { siteTitle }
          </Typography>
          { isLoggedIn
              ?
                <Button color="inherit" onClick={() => { props.logout(); }}>Log out</Button>
              :
                null
          }
        </Toolbar>
      </AppBar>
    </div>
  );
};

Header.propTypes = {
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
  logout: PropTypes.func.isRequired,
  isLoggedIn: PropTypes.bool,
};

Header.defaultProps = {
  isLoggedIn: false,
};

const mapStateToProps = () => ({});
const mapDispatchToProps = dispatch => ({
  logout: () => dispatch(logout()),
});

export default compose(
  withStyles(styles),
  connect(mapStateToProps, mapDispatchToProps),
)(Header);
