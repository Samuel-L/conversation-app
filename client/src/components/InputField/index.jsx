import React from 'react';
import PropTypes from 'prop-types';

import { withStyles } from '@material-ui/core/styles';
import FormControl from '@material-ui/core/FormControl';
import InputLabel from '@material-ui/core/InputLabel';
import Input from '@material-ui/core/Input';
import FormHelperText from '@material-ui/core/FormHelperText';

const styles = theme => ({
  formControl: {
    margin: theme.spacing.unit,
  },
});

export const InputField = (props) => {
  const { classes } = props;

  return (
    <FormControl error={props.error} className={classes.formControl}>
      <InputLabel htmlFor={props.id}>{props.label}</InputLabel>
      <Input
        id={props.id}
        type={props.type}
        value={props.value}
        onChange={props.handleChange}
      />
      <FormHelperText id={`${props.id}-help-text`}>
        { props.error ? props.errorMessage : null }
      </FormHelperText>
    </FormControl>
  );
};


InputField.propTypes = {
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
  id: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  handleChange: PropTypes.func.isRequired,
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
};

InputField.defaultProps = {
  error: false,
  errorMessage: '',
};

export default withStyles(styles)(InputField);
