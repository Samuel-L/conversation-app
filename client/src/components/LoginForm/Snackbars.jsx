import React from 'react';
import PropTypes from 'prop-types';
import HttpStatus from 'http-status-codes';

import ErrorSnack from '../ErrorSnack';
import SuccessSnack from '../SuccessSnack';

const Snackbars = props => (
  <React.Fragment>
    { props.error === HttpStatus.BAD_REQUEST
      ?
        <ErrorSnack
          message="Wrong username or password"
          error={props.error === HttpStatus.BAD_REQUEST}
          handleClose={props.handleClose}
        />
      :
        null
    }
    { props.error !== HttpStatus.BAD_REQUEST
      ?
        <ErrorSnack
          message="Something went wrong, please try again"
          error={props.error !== null}
          handleClose={props.handleClose}
        />
      :
        null
    }
    { props.success
      ?
        <SuccessSnack
          success={props.success}
          message="You are logged in."
          handleClose={props.handleClose}
        />
      :
        null
    }
  </React.Fragment>
);

Snackbars.propTypes = {
  error: PropTypes.number,
  success: PropTypes.bool,
  handleClose: PropTypes.func.isRequired,
};

Snackbars.defaultProps = {
  error: null,
  success: false,
};

export default Snackbars;
