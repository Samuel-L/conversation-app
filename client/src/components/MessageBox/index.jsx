import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import Button from '@material-ui/core/Button';

import InputField from '../InputField';
import Message from './Message';
import { sendMessage } from '../../redux-modules/send-message';

class MessageBox extends Component {
  state = {
    message: '',
  };

  handleTextChange = (event) => {
    this.setState({ [event.target.id]: event.target.value });
  };

  handleSendMessage = () => {
    const { message } = this.state;
    const { conversationID } = this.props;
    this.props.sendMessage(conversationID, message);
    this.setState({ message: '' });
  };

  render() {
    return (
      <div>
        <div>
          { this.props.messages
            ?
              this.props.messages.map(message => (
                <Message
                  key={message.id}
                  message={message}
                />
              ))
            :
              null
          }
        </div>
        <div>
          <InputField
            id="message"
            label="Message"
            type="text"
            value={this.state.message}
            handleChange={this.handleTextChange}
          />
          <Button variant="raised" onClick={this.handleSendMessage}>
            Send
          </Button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  success: state.sendMessageReducer.sent,
  failure: state.sendMessageReducer.error,
});

const mapDispatchToProps = dispatch => ({
  sendMessage: (conversationID, message) => dispatch((sendMessage(conversationID, message))),
});

MessageBox.propTypes = {
  messages: PropTypes.arrayOf(PropTypes.object),
  sendMessage: PropTypes.func.isRequired,
  conversationID: PropTypes.number,
};

MessageBox.defaultProps = {
  messages: [{}],
  conversationID: 0,
};

export default connect(mapStateToProps, mapDispatchToProps)(MessageBox);
