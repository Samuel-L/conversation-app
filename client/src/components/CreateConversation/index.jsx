import React, { Component } from 'react';
import PropTypes from 'prop-types';
import compose from 'recompose/compose';
import { connect } from 'react-redux';

import { withStyles } from '@material-ui/core/styles';
import Grid from '@material-ui/core/Grid';
import Button from '@material-ui/core/Button';
import CircularProgress from '@material-ui/core/CircularProgress';


import { createConversation, resetState } from '../../redux-modules/conversation-creator';
import Snackbars from './Snackbars';
import InputField from '../InputField';

const styles = theme => ({
  buttonGrid: {
    alignSelf: 'center',
  },
  button: {
    margin: theme.spacing.unit,
  },
});

class CreateConversation extends Component {
  state = {
    topic: '',
    counterpartUsername: '',
    errors: {
      topic: { error: false, message: '' },
      counterpartUsername: { error: false, message: '' },
    },
  };

  handleInputChange = (event) => {
    this.setState({ [event.target.id]: event.target.value });
  };

  stateFieldIsPopulated = field => (
    field !== ''
  );

  validateForm = (topic, username) => {
    const topicIsPopulated = this.stateFieldIsPopulated(topic);
    const usernameIsPopulated = this.stateFieldIsPopulated(username);

    const errors = {
      topic: { error: !topicIsPopulated, message: 'This field is required.' },
      counterpartUsername: { error: !usernameIsPopulated, message: 'This field is required.' },
    };
    this.setState({ errors });

    if (topicIsPopulated && usernameIsPopulated) {
      return true;
    }
    return false;
  };

  handleFormSubmit = (event) => {
    event.preventDefault();
    const { topic, counterpartUsername } = this.state;
    const formIsValidated = this.validateForm(topic, counterpartUsername);

    if (formIsValidated) {
      this.props.createConversation(topic, counterpartUsername);
    }
  };

  handleSnackbarClose = () => {
    this.props.resetState();
  };

  render() {
    const { classes } = this.props;
    return (
      <form onSubmit={this.handleFormSubmit}>
        <Grid container>
          <Grid item xs={12} md={5}>
            <InputField
              error={this.state.errors.topic.error}
              errorMessage={this.state.errors.topic.message}
              id="topic"
              label="Topic"
              type="text"
              value={this.state.topic}
              handleChange={this.handleInputChange}
            />
          </Grid>
          <Grid item xs={12} md={5}>
            <InputField
              error={this.state.errors.counterpartUsername.error}
              errorMessage={this.state.errors.counterpartUsername.message}
              id="counterpartUsername"
              label="Username"
              type="text"
              value={this.state.counterpartUsername}
              handleChange={this.handleInputChange}
            />
          </Grid>
          <Grid item xs={12} md={2} className={classes.buttonGrid}>
            { !this.props.creating
              ?
                <Button type="submit" variant="raised" color="primary" className={classes.button}>
                  Create
                </Button>
              :
                <CircularProgress />
            }
          </Grid>
        </Grid>
        <Snackbars
          error={this.props.error}
          success={this.props.created}
          handleClose={this.handleSnackbarClose}
        />
      </form>
    );
  }
}

CreateConversation.propTypes = {
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
  createConversation: PropTypes.func.isRequired,
  resetState: PropTypes.func.isRequired,
  creating: PropTypes.bool,
  created: PropTypes.bool,
  error: PropTypes.number,
};

CreateConversation.defaultProps = {
  creating: false,
  created: false,
  error: null,
};

const mapStateToProps = state => ({
  creating: state.conversationCreatorReducer.creating,
  created: state.conversationCreatorReducer.created,
  error: state.conversationCreatorReducer.error,
});

const mapDispatchToProps = dispatch => ({
  createConversation: (topic, username) => dispatch(createConversation(topic, username)),
  resetState: () => dispatch(resetState()),
});

export default compose(
  withStyles(styles),
  connect(mapStateToProps, mapDispatchToProps),
)(CreateConversation);
