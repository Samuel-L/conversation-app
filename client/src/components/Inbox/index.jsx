import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import Parser from '../../utils/parser';
import Item from './Item';
import Snackbars from './Snackbars';
import { fetchConversations } from '../../redux-modules/conversation-fetcher';
import { deleteConversation, resetState as resetDeleteState } from '../../redux-modules/conversation-deleter';
import { archiveConversation, resetState as resetArchiveState } from '../../redux-modules/conversation-archiver';

class Inbox extends Component {
  state = {
    expanded: null,
  };

  componentDidMount() {
    this.props.fetchConversations();
  }

  handleExpansionPanel = panel => (event, expanded) => {
    this.setState({ expanded: expanded ? panel : false });
  };

  handleItemDelete = (id) => {
    this.props.deleteConversation(id);
  };

  handleItemArchive = (id) => {
    this.props.archiveConversation(id);
  };

  handleSnackbarClose = () => {
    this.props.resetDeleteState();
    this.props.resetArchiveState();
  };

  render() {
    return (
      <div>
        {
          this.props.conversations.map(conversation => (
            <Item
              key={conversation.id}
              conversation={conversation}
              expanded={this.state.expanded}
              handleChange={this.handleExpansionPanel}
              handleDelete={this.handleItemDelete}
              handleArchive={this.handleItemArchive}
            />
          ))
        }
        <Snackbars
          handleClose={this.handleSnackbarClose}
          deleteSuccess={this.props.deleted}
          archiveSuccess={this.props.archived}
        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  conversations: state.conversationFetcherReducer.conversations,
  deleted: state.conversationDeleterReducer.deleted,
  archived: state.conversationArchiverReducer.archived,
});

const mapDispatchToProps = dispatch => ({
  fetchConversations: () => dispatch(fetchConversations()),
  deleteConversation: id => dispatch(deleteConversation(id)),
  archiveConversation: id => dispatch(archiveConversation(id)),
  resetArchiveState: () => dispatch(resetArchiveState()),
  resetDeleteState: () => dispatch(resetDeleteState()),
});

Inbox.propTypes = {
  fetchConversations: PropTypes.func.isRequired,
  deleteConversation: PropTypes.func.isRequired,
  resetDeleteState: PropTypes.func.isRequired,
  archiveConversation: PropTypes.func.isRequired,
  resetArchiveState: PropTypes.func.isRequired,
  conversations: PropTypes.arrayOf(PropTypes.instanceOf(Parser)),
  deleted: PropTypes.bool,
};

Inbox.defaultProps = {
  conversations: [new Parser(null, null)],
  deleted: false,
};

export default connect(mapStateToProps, mapDispatchToProps)(Inbox);
