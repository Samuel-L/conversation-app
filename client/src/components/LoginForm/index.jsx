import React, { Component } from 'react';
import PropTypes from 'prop-types';
import compose from 'recompose/compose';
import { connect } from 'react-redux';

import { withStyles } from '@material-ui/core/styles';
import Grid from '@material-ui/core/Grid';
import Button from '@material-ui/core/Button';
import CircularProgress from '@material-ui/core/CircularProgress';

import InputField from '../InputField';
import Snackbars from './Snackbars';
import { login, resetState } from '../../redux-modules/auth';

const styles = theme => ({
  button: {
    margin: theme.spacing.unit,
  },
});

export class LoginForm extends Component {
  state = {
    username: '',
    password: '',
    errors: {
      username: { error: false, message: '' },
      password: { error: false, message: '' },
    },
  };

  stateFieldIsPopulated = field => (
    field !== ''
  );

  validateForm = (username, password) => {
    const usernameIsPopulated = this.stateFieldIsPopulated(username);
    const passwordIsPopulated = this.stateFieldIsPopulated(password);
    const errors = {
      username: { error: !usernameIsPopulated, message: 'This field is required.' },
      password: { error: !passwordIsPopulated, message: 'This field is required.' },
    };
    this.setState({ errors });

    if (usernameIsPopulated && passwordIsPopulated) {
      return true;
    }
    return false;
  };

  handleFormSubmit = (event) => {
    event.preventDefault();
    const { username, password } = this.state;
    const formIsValidated = this.validateForm(username, password);

    if (formIsValidated) {
      this.props.login(username, password);
    }
  };

  handleInputChange = (event) => {
    this.setState({ [event.target.id]: event.target.value });
  };

  handleSnackbarClose = () => {
    this.props.resetState();
  };

  render() {
    const { classes } = this.props;

    return (
      <form onSubmit={this.handleFormSubmit}>
        <Grid container>
          <Grid item xs={12}>
            <InputField
              error={this.state.errors.username.error}
              errorMessage={this.state.errors.username.message}
              id="username"
              label="Username"
              type="text"
              value={this.state.username}
              handleChange={this.handleInputChange}
            />
          </Grid>
          <Grid item xs={12}>
            <InputField
              error={this.state.errors.password.error}
              errorMessage={this.state.errors.password.message}
              id="password"
              label="Password"
              type="password"
              value={this.state.password}
              handleChange={this.handleInputChange}
            />
          </Grid>
          <Grid item xs={12}>
            { !this.props.loading
              ?
                <Button type="submit" variant="raised" color="primary" className={classes.button}>
                  Login
                </Button>
              :
                <CircularProgress />
            }
          </Grid>
        </Grid>
        <Snackbars
          error={this.props.error}
          success={this.props.success}
          handleClose={this.handleSnackbarClose}
        />
      </form>
    );
  }
}

LoginForm.propTypes = {
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
  login: PropTypes.func.isRequired,
  resetState: PropTypes.func.isRequired,
  loading: PropTypes.bool,
  success: PropTypes.bool,
  error: PropTypes.number,
};

LoginForm.defaultProps = {
  loading: false,
  success: false,
  error: null,
};

const mapStateToProps = state => ({
  loading: state.authReducer.loggingIn,
  success: state.authReducer.loggedIn,
  error: state.authReducer.error,
});

const mapDispatchToProps = dispatch => ({
  login: (username, password) => dispatch(login(username, password)),
  resetState: () => dispatch(resetState()),
});

export default compose(
  withStyles(styles),
  connect(mapStateToProps, mapDispatchToProps),
)(LoginForm);
