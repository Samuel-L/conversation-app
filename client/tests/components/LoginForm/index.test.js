import React from 'react';
import renderer from 'react-test-renderer';
import { createShallow } from '@material-ui/core/test-utils';

import { LoginForm } from '../../../src/components/LoginForm';

describe('component: LoginForm', () => {
  const login = jest.fn();
  const resetState = jest.fn();
  const handleChangeComponent = jest.fn();
  const classes = { button: '' };
  let wrapper;

  beforeEach(() => {
    const shallow = createShallow();
    wrapper = shallow(
      <LoginForm
        login={login}
        resetState={resetState}
        handleChangeComponent={handleChangeComponent}
        classes={classes}
      />
    );
  });

  describe('rendering', () => {
    it('matches snapshot', () => {
      const tree = renderer.create(
        <LoginForm
          login={login}
          resetState={resetState}
          classes={classes}
          handleChangeComponent={handleChangeComponent}
        />
      ).toJSON(); 

      expect(tree).toMatchSnapshot();
    });
  });

  describe('functionality', () => {
    it('takes username input', () => {
      const usernameInput = wrapper.find('#username').dive().dive().find('#username');
      usernameInput.simulate('change', { target: { id: 'username', value: 'Username' } });

      expect(wrapper.state('username')).toEqual('Username');
    });

    it('takes password input', () => {
      const passwordInput = wrapper.find('#password').dive().dive().find('#password');
      passwordInput.simulate('change', { target: { id: 'password', value: 'Password' } });

      expect(wrapper.state('password')).toEqual('Password');
    });
  });
});
