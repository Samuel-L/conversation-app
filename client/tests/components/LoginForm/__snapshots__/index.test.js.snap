// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: LoginForm rendering matches snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="MuiGrid-container-1"
  >
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-12-39"
    >
      <div
        className="MuiFormControl-root-93 InputField-formControl-92"
        onBlur={[Function]}
        onFocus={[Function]}
      >
        <label
          className="MuiFormLabel-root-102 MuiInputLabel-root-97 MuiInputLabel-formControl-98 MuiInputLabel-animated-101"
          data-shrink={false}
          htmlFor="username"
        >
          Username
        </label>
        <div
          className="MuiInput-root-107 MuiInput-formControl-108 MuiInput-underline-111"
        >
          <input
            aria-invalid={false}
            aria-required={false}
            className="MuiInput-input-115"
            disabled={false}
            id="username"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="text"
            value=""
          />
        </div>
        <p
          className="MuiFormHelperText-root-120"
          id="username-help-text"
        />
      </div>
    </div>
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-12-39"
    >
      <div
        className="MuiFormControl-root-93 InputField-formControl-92"
        onBlur={[Function]}
        onFocus={[Function]}
      >
        <label
          className="MuiFormLabel-root-102 MuiInputLabel-root-97 MuiInputLabel-formControl-98 MuiInputLabel-animated-101"
          data-shrink={false}
          htmlFor="password"
        >
          Password
        </label>
        <div
          className="MuiInput-root-107 MuiInput-formControl-108 MuiInput-underline-111"
        >
          <input
            aria-invalid={false}
            aria-required={false}
            className="MuiInput-input-115 MuiInput-inputType-118"
            disabled={false}
            id="password"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="password"
            value=""
          />
        </div>
        <p
          className="MuiFormHelperText-root-120"
          id="password-help-text"
        />
      </div>
    </div>
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-12-39"
    >
      <button
        className="MuiButtonBase-root-146 MuiButton-root-124 MuiButton-contained-133 MuiButton-containedPrimary-134 MuiButton-raised-136 MuiButton-raisedPrimary-137"
        disabled={false}
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="submit"
      >
        <span
          className="MuiButton-label-125"
        >
          Login
        </span>
        <span
          className="MuiTouchRipple-root-149"
        />
      </button>
    </div>
  </div>
</form>
`;
